daemon/container.go

```diff
@@ -256,14 +256,44 @@ func populateCommand(c *Container, env []string) error {
 	pid.HostPid = c.hostConfig.PidMode.IsHost()
 
 	// Build lists of devices allowed and created within the container.
-	userSpecifiedDevices := make([]*devices.Device, len(c.hostConfig.Devices))
-	for i, deviceMapping := range c.hostConfig.Devices {
+	var userSpecifiedDevices []*devices.Device
+	for _, deviceMapping := range c.hostConfig.Devices {
 		device, err := devices.GetDevice(deviceMapping.PathOnHost, deviceMapping.CgroupPermissions)
+		// if the device is not a device node
+		// let's try to see if it's a directory holding many devices
+		if err == devices.ErrNotADeviceNode {
+			// check if it is a directory
+			if src, e := os.Stat(deviceMapping.PathOnHost); e == nil && src.IsDir() {
+				// mount the internal devices recursively
+				var foundChildDevice bool
+				filepath.Walk(deviceMapping.PathOnHost, func(dpath string, f os.FileInfo, e error) error {
```

Comment:
I wonder if it is possible to make only one `filepath.Walk` because now I am confused with code flow :)